src/jsonrpc.c

```diff
@@ -0,0 +1,493 @@
+
+		if (body_len > SSIZE_MAX - chunk_len) {
+			jsonrpc_log(req, LOG_CRIT,
+			    "Request body bigger than the platform accepts");
+			return (JSONRPC_SERVER_ERROR);
+		}
+		body_len += chunk_len;
+
+		if (body_len > body_max_len) {
+			jsonrpc_log(req, LOG_ERR,
+			    "Request overreached configured body size limit");
+			return (JSONRPC_LIMIT_REACHED);
+		}
+		kore_buf_append(&req->buf, chunk_buffer, chunk_len);
+	}
+
+	/* Grab our body data as a NUL-terminated string. */
+	body_string = kore_buf_stringify(&req->buf, NULL) + body_start;
+
+	/* Parse the body via yajl now. */
+	*error_buffer = 0;
+	req->json = yajl_tree_parse(body_string, error_buffer,
+	    sizeof(error_buffer));
+	if (req->json == NULL) {
+		if (strlen(error_buffer)) {
+			jsonrpc_log(req, LOG_ERR, "Invalid json: %s",
+			    error_buffer);
+		} else {
+			jsonrpc_log(req, LOG_ERR, "Invalid json");
+		}
+		return (JSONRPC_PARSE_ERROR);
+	}
+
+	return (0);
+}
+
+static int
+parse_json_body(struct jsonrpc_request *req)
+{
+	static const char	*proto_path[] = { "jsonrpc", NULL };
+	static const char	*id_path[] = { "id", NULL };
+	static const char	*method_path[] = { "method", NULL }; 
+	static const char	*params_path[] = { "params", NULL };
+
+	/* Check protocol first. */
+	yajl_val proto = yajl_tree_get(req->json, proto_path, yajl_t_string);
+	if (proto == NULL) {
+		jsonrpc_log(req, LOG_ERR,
+		    "JSON-RPC protocol MUST be indicated and \"2.0\"");
+		return (JSONRPC_PARSE_ERROR);
+	}
+
+	char *proto_string = YAJL_GET_STRING(proto);
+	if (proto_string == NULL) {
+		jsonrpc_log(req, LOG_ERR,
+		    "JSON-RPC protocol MUST be indicated and \"2.0\"");
+		return (JSONRPC_PARSE_ERROR);
+	}
+
+	if (strcmp("2.0", proto_string) != 0) {
+		jsonrpc_log(req, LOG_ERR,
+		    "JSON-RPC protocol MUST be indicated and \"2.0\"");
+		return (JSONRPC_PARSE_ERROR);
+	}
+
+	/* Check id. */ 
+	if ((req->id = yajl_tree_get(req->json, id_path, yajl_t_any)) != NULL) {
+		if (YAJL_IS_NUMBER(req->id)) {
+			if (!YAJL_IS_INTEGER(req->id)) {
+				jsonrpc_log(req, LOG_ERR,
+				    "JSON-RPC id SHOULD NOT contain fractional"
+				    " parts");
+				return (JSONRPC_PARSE_ERROR);
+			}
+		} else if (!YAJL_IS_STRING(req->id)) {
+			jsonrpc_log(req, LOG_ERR,
+			    "JSON-RPC id MUST contain a String or Number");
+			return (JSONRPC_PARSE_ERROR);
+		}
+	}
+
+	/* Check method. */
+	if ((req->method = YAJL_GET_STRING(yajl_tree_get(req->json, method_path,
+		yajl_t_string))) == NULL) {
+		jsonrpc_log(req, LOG_ERR,
+		    "JSON-RPC method MUST exist and be a String");
+		return (JSONRPC_PARSE_ERROR);
+	}
+
+	/* Check params. */
+	req->params = yajl_tree_get(req->json, params_path, yajl_t_any);
+	if (!(req->params == NULL || YAJL_IS_ARRAY(req->params)
+	    || YAJL_IS_OBJECT(req->params))) {
+		jsonrpc_log(req, LOG_ERR,
+		    "JSON-RPC params MUST be Object or Array");
+		return (JSONRPC_PARSE_ERROR);
+	}
+
+	return (0);
+}
+
+int
+jsonrpc_request_read(struct http_request *http_req, ssize_t max_body_len,
```

Comment:
This function is going to return JSONRPC specific return values. This is wrong.

It should return KORE_RESULT_OK or KORE_RESULT_ERROR.

Please store any error codes related to JSONRPC in the jsonrpc_request data structure instead.